### PR TITLE
Move ``misplaced-comparison-constant`` to optional extension

### LINKED ISSUE #1064: Disable 'misplaced-comparison-constant' checker by default by moving it to extensions
### Steps to reproduce
1.  Write a comparison such as `if 5 < a:...`
### Current behavior

Pylint complains with 'misplaced-comparison-constant'.
### Expected behavior

No issue.

In other languages, so-called ["Yoda expressions"](https://en.wikipedia.org/wiki/Yoda_conditions) can be used as a code guard to prevent an accident assignment when an equality comparison was indented. (i.e. `if (42 = x) { /* in C, this causes a compile time error error */ }` In Python, we cannot use assignments when a boolean expression is expected, so let's not dwell on this use case any further.

However, that is not my reason for using this style of comparison. I prefer to avoid using the greater-than symbol (`>`). Instead, I'll always use the less-than (`<`) or the less-than-equal (`<=`). I'll yield to Dijkstra's explanation for the best way to describe ranges in his paper ['Why numbering should start at zero'](https://www.cs.utexas.edu/users/EWD/transcriptions/EWD08xx/EWD831.html).

Given we are talking Python, that paper is, actually, very interesting because Python is one of the very few languages that the range expression is a valid boolean expression; i.e. `2 <= x < 13`. So, I'll cynically ask, "Is that half-good, half-bad?"

A quick test against pylint tells me that the range expression is handled appropriately.

So, why do we even care?

Readability!? If so, that seems inane to me. Force the nitpickers to enable `misplaced-comparison-constant` rather than ding those that aren't so easily confused.

Other rationale!? I can't think of some use case that may invoke some rare peculiarity as one may find with overloaded operators in C++. That is to say, Python will use reflection to handle equality of unlike objects; assuming the objects make use of `NotImplemented`.

Example:

``` python
"""This example is intentionally absurd."""

class MyClass(object):
    def __eq__(self, other):
        if isintance(other, int):
            return other == 42
        return NotImplemented

class MyOtherClass(object):
    def __eq__(self, other):
        return isinstance(other, MyClass)

MyClass() == 42  # True
42 == MyClass()  # True

MyOtherClass() == MyClass()  # True
MyClass() == MyOtherClass()  # True
```
### pylint --version output

```
$ pylint --version
No config file found, using default configuration
pylint 1.6.4,
astroid 1.4.7
Python 2.7.11 (default, Mar 15 2016, 09:47:58)
[GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)]
```

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
| ✓   | :sparkles: New feature |
| ✓   | :hammer: Refactoring   |

## Description

Closes #1064

Consensus was reached that this check should be an extension. I thought about putting it in `code_style`, but since that only has `Recommendation` and this is a `Convention` I thought I might as well create a new extension.
Since the extensions loaded in our `pylintrc` are not automatically loaded in the functional tests this required to update some tests and their results.

### CODE REVIEW

**Comment by DanielNoord** [comment]
The messages that got removed from `CompareToZeroUsedTC` are covered by the functional test. I don't think we need to add a unittest for this extension at this moment.

File: pylint/extensions/comparison_placement.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+
+"""Checks the placement of comparisons"""
```
Comment (Pierre-Sassoulas) [inline pylint/extensions/comparison_placement.py:4]:
```suggestion
"""
Checks for yoda comparisons (variable before constant)
See https://en.wikipedia.org/wiki/Yoda_conditions
"""
```
I think it's ok to be explicit and use colloquialism

File: pylint/extensions/comparison_placement.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+from astroid import nodes
+
+from pylint.checkers import BaseChecker, utils
+from pylint.interfaces import IAstroidChecker
+
+REVERSED_COMPS = {"<": ">", "<=": ">=", ">": "<", ">=": "<="}
+COMPARISON_OPERATORS = frozenset(("==", "!=", "<", ">", "<=", ">="))
+
+
+class ComparisonPlacementChecker(BaseChecker):
```
Comment (Pierre-Sassoulas) [inline pylint/extensions/comparison_placement.py:16]: Is this a checker that is supposed to stay with one message forever, we could use the same name than the message if that's the case ?
```suggestion
class MisplacedComparisonConstantChecker(BaseChecker):
```

File: pylint/extensions/comparison_placement.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    """Checks the placement of constants in comparisons"""
```
Comment (Pierre-Sassoulas) [inline pylint/extensions/comparison_placement.py:17]: It feels like it's duplicated from the module docstring, we might as well remove one of them.

**Review by Pierre-Sassoulas — COMMENTED** [review]
Some suggestions otherwise  LGTM :) Thank you for handling the work for a decision taken a long time ago.

File: pylint/extensions/comparison_placement.py (hunk 3 above)
Comment (DanielNoord) [reply pylint/extensions/comparison_placement.py:17]: We would miss out on IntelliSense on the class then. I also saw that we are waiting to add `extensions.docparams` in our `pylintrc` which would requires us to re-add this docstring in the future.

**Comment by DanielNoord** [comment]
Updated with additional tests to make sure coverage doesn't decrease. Ready for review again!

File: pylint/extensions/comparison_placement.py (hunk 3 above)
Comment (Pierre-Sassoulas) [reply pylint/extensions/comparison_placement.py:20]: Fair enough.

File: tests/functional/m/misplaced_comparison_constant.py [hunk 4]
```diff
@@ -34,3 +34,14 @@ def good_comparison():
     for i in range(10):
         if i == 5:
             pass
+
+def double_comparison():
+    """Check that we return early for non-binary comparison"""
+    for i in range(10):
+        if i == 1 == 2:
+            pass
```
Comment (Pierre-Sassoulas) [inline tests/functional/m/misplaced_comparison_constant.py:42]:
```suggestion
        if i == 1 == 2:
            pass
        if 2 <= i <= 8:
            print("Between 2 and 8 inclusive")
```
I think adding the standard use case could be safe as the other one is kinda theoretical.

**Review by Pierre-Sassoulas — APPROVED** [review]
Little nitpick to add the standard test case for double comparison in the functional tests, it's still good to go.

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas Turns out the CI on `main` has been failing since this PR. Not sure if you noticed, just thought I would let you know.

**Comment by Pierre-Sassoulas** [comment]
Ha it look like it's because of my experiment with Github pages. Now I don't even know how to remove it :(

**Pull request merged**